### PR TITLE
Build OpenMC as a library and pass MPI intracommunicator

### PR DESCRIPTION
**Pull request by paulromano**
This pull request restructures OpenMC to look a little more like a library. Most importantly, the `openmc_init` subroutine (renamed from `initialize_run` to avoid namespace issues) now accepts an MPI intracommunicator as an argument. These changes will both 1) allow an external application to call and control OpenMC and 2) to indicate to OpenMC to only operate on a subset of MPI processes if need be. The CMake file now builds all files other than main.F90 as a static library (libopenmc).

For those interested -- these changes are to support TH coupling with Nek5000.

### CODE REVIEW

File: src/finalize.F90 [hunk 1]
```diff
@@ -1,17 +1,14 @@
 module finalize
 
+  use hdf5, only: h5tclose_f, h5close_f
+
```
Comment (wbinventor) [inline src/finalize.F90]: Is this newline here for a reason? Any reason to split up the headers from `hdf5` and `hdf5_interface`?

File: src/initialize.F90 [hunk 2]
```diff
@@ -27,38 +35,31 @@ module initialize
...
+! OPENMC_INIT takes care of all initialization tasks, i.e. reading
 ! from command line, reading xml input files, initializing random
 ! number seeds, reading cross sections, initializing starting source,
 ! setting up timers, etc.
 !===============================================================================
 
-  subroutine initialize_run()
+  subroutine openmc_init(intracomm)
+#ifdef MPIF08
+    type(MPI_Comm), intent(in) :: intracomm
```
Comment (wbinventor) [inline src/initialize.F90]: Perhaps comment on the `intracomm` parameter

File: src/initialize.F90 [hunk 3]
```diff
@@ -164,7 +165,12 @@ end subroutine initialize_run
 ! each processor.
 !===============================================================================
 
-  subroutine initialize_mpi()
+  subroutine initialize_mpi(intracomm)
+#ifdef MPIF08
+    type(MPI_Comm), intent(in) :: intracomm
```
Comment (wbinventor) [inline src/initialize.F90]: ditto

File: CMakeLists.txt [hunk 4]
```diff
@@ -266,15 +266,105 @@ add_library(faddeeva STATIC src/Faddeeva.c)
 #===============================================================================
 
 set(program "openmc")
-file(GLOB source src/*.F90 src/xml/openmc_fox.F90)
-add_executable(${program} ${source})
+set(LIBOPENMC_FORTRAN_SRC
```
Comment (paulromano) [inline CMakeLists.txt]: It's actually a recommendation to [not use `GLOB` to list source files](https://cmake.org/cmake/help/latest/command/file.html?highlight=glob#file).

File: src/finalize.F90 (hunk 1 above)
Comment (paulromano) [inline src/finalize.F90]: Yes, to separate external dependencies (HDF5) from internal.

**Pull request merged**